tests/algorithms/test_gradient_clipping.py

```diff
@@ -58,56 +58,122 @@ def forward(self, x):
+    new_gc_fn.assert_called_once()
 
+    new_gc_fn = Mock()
+    monkeypatch.setattr(torch.nn.utils, 'clip_grad_value_', new_gc_fn)
+    apply_gradient_clipping(parameters, 'value', 0.1)
+    new_gc_fn.assert_called_once()
 
-def test_agc_functional_with_cnn_does_not_error(cnn_model_with_grads):
-    """This test is just to ensure that no errors are raised.
 
-    Accuracy of the AGC calculations are tested in other tests.
-    """
+@pytest.mark.parametrize('clipping_type', [('adaptive',), ('norm',), ('value',)])
+def test_gradient_clipping_algorithm(monkeypatch, clipping_type, simple_model_with_grads):
+    model = simple_model_with_grads
+    apply_gc_fn = Mock()
+    monkeypatch.setattr(gc_module, 'apply_gradient_clipping', apply_gc_fn)
+    state = Mock()
+    state.is_model_deepspeed = False
+    state.model = model
+    state.profiler.marker = Mock(return_value=None)
+    state.callbacks = []
+    state.algorithms = [GradientClipping(clipping_type=clipping_type, clipping_threshold=0.01)]
+    logger = Mock()
+    engine = Engine(state, logger)
 
-    model = cnn_model_with_grads
-    # Call apply_agc. If this function returns then we know that nothing errored out.
-    cf.apply_agc(model)
+    # Run the Event that should cause AGC.apply to be called.
+    engine.run_event(Event.AFTER_TRAIN_BATCH)
 
+    apply_gc_fn.assert_called_once()
 
-def test_AGC_algorithm(simple_model_with_grads):
-    # Get weight and gradients from the fixture model.
-    model = simple_model_with_grads
-    weights = next(model.parameters())
-    grad = weights.grad
-    expected_clipped_grad = grad.detach() * _get_clipped_gradient_coeff(weights, grad)
 
-    # Set up a mock engine.
+def test_gradient_clipping_algorithm_with_deepspeed_enabled(monkeypatch: pytest.MonkeyPatch):
+    clipping_threshold = 0.1191
+    apply_gc_fn = Mock()
+    monkeypatch.setattr(gc_module, 'apply_gradient_clipping', apply_gc_fn)
     state = Mock()
-    state.model = model
     state.profiler.marker = Mock(return_value=None)
     state.callbacks = []
-    state.algorithms = [AGC()]
+
+    # Enable deepspeed and set clipping_type to norm to ensure that apply_gradient_clipping
+    # is not called.
+    state.algorithms = [GradientClipping(clipping_type='norm', clipping_threshold=clipping_threshold)]
+    state.is_model_deepspeed = True
+
+    state.optimizers = Mock()
+    state.optimizers.clip_grad = 0.0
     logger = Mock()
     engine = Engine(state, logger)
 
     # Run the Event that should cause AGC.apply to be called.
     engine.run_event(Event.AFTER_TRAIN_BATCH)
 
-    # Check that the gradients weights holds are equivalent to the calling
-    # grad.detach() * _get_clipped_gradient_coeff on the weights and grads.
+    # Make sure deepspeed engine's gradient_clipping_field is set properly.
+    assert state.optimizers.clip_grad == clipping_threshold
+    # Make sure apply_gradient_clipping is not called.
+    apply_gc_fn.assert_not_called()
```

Comment:
I don't think this tests that gradients were actually clipped when using deepspeed. Would it be possible to inspect the optimizer weights?

Reply:
now with setting deepspeed_config here: [cc8830d](https://github.com/mosaicml/composer/pull/1099/commits/cc8830dc422f9a472fc61fba673f07ba9e2d22b3), I don't think it's necessary to check the gradients are clipped if we trust the DeepSpeed API